flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/adaptive/AdaptiveJoinOperatorGeneratorTest.java

```diff
@@ -0,0 +1,424 @@
+                new UniformBinaryRowGenerator(numKeys1, buildValsPerKey, true);
+        MutableObjectIterator<BinaryRowData> probeInput =
+                new UniformBinaryRowGenerator(numKeys2, probeValsPerKey, true);
+
+        buildJoin(
+                buildInput,
+                probeInput,
+                originalJoinType,
+                false,
+                true,
+                isBuildLeft,
+                isBroadcast,
+                isBuildLeft ? 280 : 270,
+                numKeys2,
+                -1);
+    }
+
+    private void testSemiJoin(OperatorType originalJoinType, boolean isBroadcast) throws Exception {
+        int numKeys1 = 9;
+        int numKeys2 = 10;
+        int buildValsPerKey = 3;
+        int probeValsPerKey = 10;
+        if (originalJoinType == OperatorType.SortMergeJoin && !isBroadcast) {
+            numKeys1 = 10;
+            numKeys2 = 9;
+            buildValsPerKey = 10;
+            probeValsPerKey = 3;
+        }
+        MutableObjectIterator<BinaryRowData> buildInput =
+                new UniformBinaryRowGenerator(numKeys1, buildValsPerKey, true);
+        MutableObjectIterator<BinaryRowData> probeInput =
+                new UniformBinaryRowGenerator(numKeys2, probeValsPerKey, true);
+
+        Object operator = newOperator(FlinkJoinType.SEMI, false, isBroadcast, originalJoinType);
+        joinAndAssert(operator, buildInput, probeInput, 90, 9, 45, true);
+    }
+
+    private void testAntiJoin(OperatorType originalJoinType, boolean isBroadcast) throws Exception {
+        int numKeys1 = 9;
+        int numKeys2 = 10;
+        int buildValsPerKey = 3;
+        int probeValsPerKey = 10;
+        if (originalJoinType == OperatorType.SortMergeJoin && !isBroadcast) {
+            numKeys1 = 10;
+            numKeys2 = 9;
+            buildValsPerKey = 10;
+            probeValsPerKey = 3;
+        }
+        MutableObjectIterator<BinaryRowData> buildInput =
+                new UniformBinaryRowGenerator(numKeys1, buildValsPerKey, true);
+        MutableObjectIterator<BinaryRowData> probeInput =
+                new UniformBinaryRowGenerator(numKeys2, probeValsPerKey, true);
+
+        Object operator = newOperator(FlinkJoinType.ANTI, false, isBroadcast, originalJoinType);
+        joinAndAssert(operator, buildInput, probeInput, 10, 1, 45, true);
+    }
+
+    public void buildJoin(
+            MutableObjectIterator<BinaryRowData> buildInput,
+            MutableObjectIterator<BinaryRowData> probeInput,
+            OperatorType originalJoinType,
+            boolean leftOut,
+            boolean rightOut,
+            boolean buildLeft,
+            boolean isBroadcast,
+            int expectOutSize,
+            int expectOutKeySize,
+            int expectOutVal)
+            throws Exception {
+        FlinkJoinType flinkJoinType = getJoinType(leftOut, rightOut);
+        Object operator = newOperator(flinkJoinType, buildLeft, isBroadcast, originalJoinType);
```

Comment:
Can we add assertion on the type of this operator?
From the current test one can hardly tell wether the operator is converted into a `BroadcastHashJoin` or not.